src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.RenameNamespaceCodeAction.cs

```diff
@@ -0,0 +1,295 @@
+            {
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+                var root = editor.OriginalRoot;
+                var containers = new HashSet<SyntaxNode>();
+
+                foreach (var refLoc in refLocations)
+                {
+                    Debug.Assert(document.Id == refLoc.Document.Id);
+
+                    // Ignore references via alias. For simple cases where the alias is defined as the type we are interested,
+                    // it will be handled properly because it is one of the reference to the type symbol. Otherwise, we don't
+                    // attempt to make a potential fix, and user might end up with errors as a result.                    
+                    if (refLoc.Alias != null)
+                    {
+                        continue;
+                    }
+
+                    // Other documents in the solution might have changed after we calculated those ReferenceLocation, 
+                    // so we can't trust anything to be still up-to-date except their spans.
+
+                    // Get inner most node in case of type used as a base type. e.g.
+                    //
+                    //      public class Foo {}
+                    //      public class Bar : Foo {}
+                    //
+                    // For the reference to Foo where it is used as a base class, the BaseTypeSyntax and the TypeSyntax
+                    // have exact same span.
+
+                    var refNode = root.FindNode(refLoc.Location.SourceSpan, getInnermostNodeForTie: true);
+                    if (_service.TryGetReplacementSyntax(refNode, namespaceParts, out var oldNode, out var newNode))
+                    {
+                        newNode = newNode.WithAdditionalAnnotations(Simplifier.Annotation);
+                        editor.ReplaceNode(oldNode, newNode);
+                    }
+
+                    var container = addImportService.GetImportContainer(root, refNode, _service.CreateUsingDirective(ImmutableArray.Create("Dummy")));
+                    containers.Add(container);
+                }
+                return (editor.GetChangedDocument(), containers.Select(c => root.GetCurrentNode(c)).ToImmutableHashSet());
+            }            
+
+            private async Task<Document> RemoveUnnecessaryImportsAsync(
+                Document document, 
+                IEnumerable<SyntaxNode> importsToRemove, 
+                DocumentOptionSet optionSet, 
+                CancellationToken cancellationToken)
+            {
+                var removeImportService = document.GetLanguageService<IRemoveUnnecessaryImportsService>();
+                return await removeImportService.RemoveUnnecessaryImportsFromCurrentContextAsync(
+                    document,
+                    import => importsToRemove.Any(importToRemove => importToRemove.IsEquivalentTo(import, topLevel: false)),
```

Comment:
1. probably better to use ISyntaxFacts.AreEquivalent so you're whitepsace insensitive.
2. how does this code differ from VB?
3. no need for await or ConfigureAwait(false)

Reply:
ignore '2', this is in shared code.